crates/ruff_linter/src/suppression.rs

```diff
@@ -159,81 +163,156 @@ impl Suppressions {
-                        codes: Some(codes),
+                        codes: Some(UnusedCodes::default()),
                         kind: UnusedNOQAKind::Suppression,
                     },
-                    range,
+                    error.range,
                 );
-                diagnostic.set_fix(Fix::safe_edit(edit));
+                diagnostic.set_fix(Fix::safe_edit(delete_comment(error.range, locator)));
             }
         }
 
-        for error in self
-            .errors
-            .iter()
-            .filter(|error| error.kind == ParseErrorKind::MissingCodes)
-        {
-            let mut diagnostic = context.report_diagnostic(
-                UnusedNOQA {
-                    codes: Some(UnusedCodes::default()),
-                    kind: UnusedNOQAKind::Suppression,
-                },
-                error.range,
-            );
-            diagnostic.set_fix(Fix::safe_edit(delete_comment(error.range, locator)));
+        if context.is_rule_enabled(Rule::InvalidRuleCode) {
+            for suppression in self
+                .valid
+                .iter()
+                .filter(|suppression| Rule::from_code(&suppression.code).is_err())
+            {
+                for comment in &suppression.comments {
+                    let (range, edit) =
+                        Suppressions::delete_code_or_comment(locator, suppression, comment);
+                    let mut diagnostic = context.report_diagnostic(
+                        InvalidRuleCode {
+                            rule_code: suppression.code.to_string(),
+                        },
+                        range,
+                    );
+                    diagnostic.set_fix(Fix::safe_edit(edit));
+                }
+            }
         }
+
+        if context.is_rule_enabled(Rule::InvalidSuppressionComment) {
+            // missing codes already handled above, report the rest as invalid comments
+            for error in self
+                .errors
+                .iter()
+                .filter(|error| error.kind != ParseErrorKind::MissingCodes)
+            {
+                let mut diagnostic = context.report_diagnostic(
+                    InvalidSuppressionComment {
+                        kind: InvalidSuppressionCommentKind::Error(error.kind),
+                    },
+                    error.range,
+                );
+                diagnostic.set_fix(Fix::safe_edit(delete_comment(error.range, locator)));
+            }
+
+            for invalid in &self.invalid {
+                let mut diagnostic = context.report_diagnostic(
+                    InvalidSuppressionComment {
+                        kind: InvalidSuppressionCommentKind::Invalid(invalid.kind),
+                    },
+                    invalid.comment.range,
+                );
+                diagnostic.set_fix(Fix::safe_edit(delete_comment(
+                    invalid.comment.range,
+                    locator,
+                )));
```

Comment:
I'm not sure we should offer a fix for invalid suppression comments. At least not a safe fix. We don't really know what the user wants to do in this case. Fix the comment? Remove it?